pkg/receive/handler.go

```diff
@@ -481,6 +505,21 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, replicas ma
 				})
 			})
 			if err != nil {
+				// Check if peer connection is unavailable, don't attempt to send requests constantly.
+				if st, ok := status.FromError(err); ok {
+					if st.Code() == codes.Unavailable {
+						h.mtx.Lock()
+						if b, ok := h.backoffs[endpoint]; ok {
+							b.attempt++
+							dur := h.expBackoff.ForAttempt(b.attempt)
+							b.nextAllowed = time.Now().Add(dur)
+							level.Debug(h.logger).Log("msg", "target unavailable backing off", "for", dur)
+						} else {
+							h.backoffs[endpoint] = &peerState{nextAllowed: time.Now().Add(h.expBackoff.ForAttempt(0))}
+						}
+						h.mtx.Unlock()
+					}
```

Comment:
I think we could delete(endpoint) from it for cleannes on ok state, no?

Reply:
Added the delete logic in the success case. Also, we clear the states when the hashring get updated.